task6-3/list.c

```diff
@@ -0,0 +1,130 @@
+#include "list.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+
+#define CAPACITY 256
+
+typedef struct Node {
+    char *name;
+    char *phone;
+    struct Node *next;
+} Node;
+
+typedef struct List {
+    int length;
+    Node *head;
+    Node *tail;
+} List;
+
+
+List *initList() {
+    List *list = malloc(sizeof(List));
+    list->head = list->tail = NULL;
+    list->length = 0;
+    return list;
+}
+
+bool isEmpty(List *list) {
+    return list->head == NULL;
+}
+
+void add(List *list, char *name, char *phone) {
+    Node *newNode = malloc(sizeof(Node));
+    list->length++;
+    newNode->name = malloc(CAPACITY * sizeof(char));
+    newNode->phone = malloc(CAPACITY * sizeof(char));
+    strcpy(newNode->name, name);
+    strcpy(newNode->phone, phone);
+    newNode->next = NULL;
+    if (isEmpty(list)) {
+        list->head = newNode;
+        list->tail = newNode;
+    } else {
+        list->tail->next = newNode;
+        list->tail = newNode;
+    }
+}
+
+void deleteHead(List* list) {
+    if (!isEmpty(list)) {
+        list->length--;
+        Node *tempHead = list->head;
+        list->head = list->head->next;
+        free(tempHead->phone);
+        free(tempHead->name);
+        free(tempHead);
+    }
+    if(isEmpty(list)){
+        free(list);
+    }
+}
+
+void deleteList(List *list) {
+    while (!isEmpty(list)) {
+        deleteHead(list);
+    }
+    free(list);
+}
+
+int listLen(List *list) {
+    return list->length;
+}
+
+void moveHeadToNewList(List* newList, List* oldList) {
+    add(newList, oldList->head->name, oldList->head->phone);
+    deleteHead(oldList);
+}
+
+List *mergeTwoListsByKey(List *first, List *second, enum SortingKey sortingKey) {
+    List *mergedList = initList();
+
+    while (listLen(first) != 0 && listLen(second) != 0) {
+        int comp;
+        if (sortingKey == PHONE) {
+            comp = strcmp(first->head->phone, second->head->phone);
+        } else {
+            comp = strcmp(first->head->name, second->head->name);
+        }
+
+        if (comp < 0) {
+            moveHeadToNewList(mergedList, first);
+        }
+        else{
+            moveHeadToNewList(mergedList, second);
+        }
+    }
+    while (listLen(first) != 0){
+        moveHeadToNewList(mergedList, first);
+    }
+    while (listLen(second) != 0){
+        moveHeadToNewList(mergedList, second);
+    }
+    deleteList(first);
+    deleteList(second);
+    return mergedList;
+}
+
+List* readFromFile(char* fileName) {
```

Comment:
```suggestion
List* readFromFile(const char* const fileName) {
```
Функция же не собирается менять имя файла. Вообще, const в Си не очень полезны, потому что генерируют только предупреждение, но всё же чем их больше, тем лучше.